tech_reports/LLMs/llms.md

```diff
@@ -73,13 +73,175 @@ Authors:
+
+#### Program Configs and Picking the Right Matmul
+Each `ttnn` operation has its own unique program config class. In general, program configs configure the op with hyperparameters that affects their functionality and performance. There are too many ops and program configs to cover in detail. We will focus on `ttnn.matmul` since it has multiple variants and it usually requires the most care.
+
+Picking a matmul variant is a key decision in optimizing a model. The choice depends on the shapes of the inputs and outputs and how the matmul fits into the rest of the model. You choose a variant by providing a specific `program_config` to `ttnn.matmul`. The following presents three matmul variants that are commonly used in LLMs.
+
+##### Matmul 2D
+Matmul 2D gets its name because it parallelizes an `(M x K) @ (K x N)` matmul over the M and N dimensions. It is useful to have this 2D parallelization when M and N are large (usually >= 256). Rule of thumb: use matmul 2D for all matmuls in prefill mode. Generally, inputs and output to matmul 2D will be interleaved in DRAM because these matmuls should be compute bound rather than memory bound and the inputs may be too large to fit in L1. NOTE: the weights can be DRAM sharded and still work with matmul 2D.
+
+The following is a description of the program config for matmul 2D.
+Given your input tensors of shape `(M x K)` and `(K x N)` and a core grid of shape `(cores_x, cores_y)`:
+
+```python
+matmul_2d_program_config = ttnn.MatmulMultiCoreReuseMultiCastProgramConfig(
+  compute_with_storage_grid_size=(cores_x, cores_y),
+  in0_block_w=1,
+  out_subblock_h=1, # Must be divisible by per_core_M
+  out_subblock_w=1, # Must be divisible by per_core_N
+  per_core_M=math.ceil(M / 32 / cores_y),  # M / TILE_HEIGHT / Grid_Size
+  per_core_N=math.ceil(N / 32 / cores_x),  # N / TILE_WIDTH / grid width
+  transpose_mcast=False,
+  fused_activation=None,
+  fuse_batch=False,
+)
+```
+Line by line, this is what the program config means.
+
+- `ttnn.MatmulMultiCoreReuseMultiCastProgramConfig`: Selects the matmul 2D variant.
+
+- `compute_with_storage_grid_size=(cores_x, cores_y)`: Determines how many cores to execute the matmul on. Note that M is parallelized over `cores_y` and N is parallelized over `cores_x`.
+
+```python
+in0_block_w=1,
+out_subblock_h=1, # Must be divisible by per_core_M
+out_subblock_w=1, # Must be divisible by per_core_N
+```
+- `in0_block_w` should divide evenly into K. Higher is better. 
+- `out_subblock_h` and `out_subblock_w` should divide evenly into M and N respectively. Higher is better. 
+  - When FP32 accumulate is enabled, `out_subblock_h * out_subblock_w` should be <= 4. Otherwise, `out_subblock_h * out_subblock_w` should be <= 8.
```

Comment:
The dest sizes are for WH; user should refer to the dest size of the device.

ideally there should be an assert if the subblock size exceeds the dest size, is there one?